backend.py

```diff
@@ -597,17 +597,63 @@ def set_robots_for_new_turn(state):
             robot.direction = Direction.N
 
 
+def sort_by_priority(item):
+    """
+    Unpack given item to robot and card and return card's priority.
+    """
+    robot, card = item
+    return card.priority
+
+
+def sort_robots_by_cards_priority(state, register):
+    """
+    Get all the active robots, sort them according to the priority of their
+    current card.
+    """
+    robot_cards = [(robot, robot.program[register])
+                   for robot in state.get_active_robots()]
+
+    robot_cards.sort(key=sort_by_priority, reverse=True)
+    return robot_cards
+
+
+def apply_register(state, register):
+    """
+    For the given register sort the robot's list according to card's priorities.
+    Apply cards effects on the sorted robots.
+    """
+    robot_cards = sort_robots_by_cards_priority(state, register)
+
+    for robot, card in robot_cards:
+        card.apply_effect(robot, state)
+
+
 def apply_all_effects(state, registers=5):
     """
-    Play the whole game: for the given number of iterations
+    Apply all game effects: for the given number of iterations
     perform robot's cards effects and tile effects on a given game state.
     At the end ressurect the inactive robots to their starting coordinates.
     registers: default iterations count is 5, can be changed for testing purposes.
     """
-    for register in range(registers):
-        for robot in state.get_active_robots():
-            current_card = robot.program[register]
-            current_card.apply_effect(robot, state)
-        apply_tile_effects(state, register)
+    _apply_cards_and_tiles_effects(state, registers)
+
     # After last register ressurect the robots to their starting coordinates.
     set_robots_for_new_turn(state)
+
+
+def _apply_cards_and_tiles_effects(state, registers):
+    """
+    Private function without ressurect mode - for testing purposes.
+    It is called within apply_all_effects. Do not call it separately.
+    """
+    for register in range(registers):
+        # try -  except was introduced for devel purposes - it may happen that
+        # robots have no card on hand and we still want to try loading the game
+        try:
+            # Check the card's priority
+            apply_register(state, register)
+
+        except IndexError:
```

Comment:
Tohle je příliš obecné, `IndexError` může vzniknout i v jiném kódu než v jakém očekáváš.

Reply:
Dalším commitem posílám fungující řešení, ale nevím, jestli tak se to má dělat. Mrkneš, prosím?

Reply:
`LookupError` vůbec nepomáhá; je [obecnější než `IndexError`](https://github.com/pyvec/cheatsheets/blob/master/exceptions/exceptions-cs.pdf). A může ho opět způsobit cokoli z `apply_register()`.

Potřebuješ výjimku, kterou určitě nezpůsobí žádný jiný kód. Ideálně si udělej vlastní:
```python
class NoCardError(LookupError):
    """Raised when a robot doesn't have a card for the given register."""
```

Reply:
Wow.

Reply:
Výjimky jsou jenom třídy :)